### PR TITLE
add contributing guidelines

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #525

As titled. Hope these guidelines could help clarify what & how to contribute to torchtitan, and make the repo more self-service.

This PR also updates some other docs in terms of formatting and removing outdated info.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -17,20 +17,72 @@ We actively welcome your pull requests.
...
+torch.backends.cudnn.deterministic = True
+torch.backends.cudnn.benchmark = False
+```
+- If a change is expected to impact computation results, loss convergence should be verified via end-to-end training on representable datasets (e.g. Llama3 models on the C4 dataset). One can refer to the example jobs reported in [performance.md](docs/performance.md) on what configs and how many steps to run.
+- The resulted loss curve should be compared with a verified baseline.
+  - 1D FSDP – Preferred, whose effectiveness can be proved by comparing with 1D DDP and single-GPU training.
+  - 2D FSDP + TP – If 1D FSDP does not suffice to make comparisons due to limited scalability. For example, this should be the baseline when experimenting with 3D parallelisms on the Llama 3.1 405B model.
+
+#### Performance
+- Memory and WPS / MFU, which are available from logging, should meet expectations. If necessary, verify the numbers on jobs spanning multiple nodes (e.g. on 64 GPUs).
```
Comment (fegin) [inline CONTRIBUTING.md]: What's the expectations for the performance? Accuracy is easy to justify (the same or the loss curve is comparable). What's the WPS / MFU expectation? For example, if a technique can reduce the memory with no better or worse WPS / MFU, the feature should still be acceptable because it allows to train larger models.

My suggestions is to clarify the expectation of WPS / MFU. If a feature is intended to address memory issue, it may not be the best to compare with the best MFU.

Comment (tianyu-l) [reply CONTRIBUTING.md]: > What's the WPS / MFU expectation? For example, if a technique can reduce the memory with no better or worse WPS / MFU, the feature should still be acceptable because it allows to train larger models.

Agree that for different techniques, the expectation should be different. This doesn't mean we can accept arbitrary WPS/MFU regression as long as there is any memory reduction. E.g. what if a technique is "supposed" to show 80% memory reduction with 10% MFU regression, but the PR only shows 40% memory reduction with 5% MFU regression. Should we accept the PR? Clearly not.

IMO the bar might need to be set case-by-case, and as long as the PR owner can justify it (whether by comparing with other similar implementations, or achieving theoretically optimal performance, etc.), we should accept. It's like a research paper should have a solid experiment section, and reviewers can make the judgement.

> My suggestions is to clarify the expectation of WPS / MFU. If a feature is intended to address memory issue, it may not be the best to compare with the best MFU.

Sounds good, let me add something.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -17,20 +17,74 @@ We actively welcome your pull requests.
...
+```
+seed = 0
+torch.manual_seed(seed)
+torch.cuda.manual_seed(seed)
+torch.backends.cudnn.deterministic = True
+torch.backends.cudnn.benchmark = False
+```
+- If a change is expected to impact computation results, loss convergence should be verified via end-to-end training on representable datasets (e.g. Llama3 models on the C4 dataset). One can refer to the example jobs reported in [performance.md](docs/performance.md) on what configs and how many steps to run.
+- The resulted loss curve should be compared with a verified baseline.
+  - 1D FSDP – Preferred, whose effectiveness can be proved by comparing with 1D DDP and single-GPU training.
```
Comment (wconstab) [inline CONTRIBUTING.md]: can be proven

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -17,20 +17,74 @@ We actively welcome your pull requests.
...
+  - 2D FSDP + TP – If 1D FSDP does not suffice to make comparisons due to limited scalability. For example, this should be the baseline when experimenting with 3D parallelisms on the Llama 3.1 405B model.
```
Comment (wconstab) [inline CONTRIBUTING.md]: 2D FSDP + TP can be used as a baseline when 1D FSDP does not suffice...

File: README.md [hunk 4]
```diff
@@ -21,12 +21,12 @@ Our guiding principles when building `torchtitan`:
...
-* [train.py](https://github.com/pytorch/torchtitan/blob/main/train.py) - the main training loop and high-level setup code
-* [torchtitan/parallelisms/parallelize_llama.py](https://github.com/pytorch/torchtitan/blob/main/torchtitan/parallelisms/parallelize_llama.py) - helpers for applying Data Parallel, Tensor Parallel, activation checkpointing, and `torch.compile` to the model
-* [torchtitan/parallelisms/pipeline_llama.py](https://github.com/pytorch/torchtitan/blob/main/torchtitan/parallelisms/pipeline_llama.py) - helpers for applying Pipeline Parallel to the model
-* [torchtitan/checkpoint.py](https://github.com/pytorch/torchtitan/blob/main/torchtitan/checkpoint.py) - utils for saving/loading distributed checkpoints
-* [torchtitan/float8.py](https://github.com/pytorch/torchtitan/blob/main/torchtitan/float8.py) - utils for applying Float8 techniques
-* [torchtitan/models/llama/model.py](https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama/model.py) - the Llama model definition (shared for Llama2 and Llama3 variants)
+* [train.py](train.py) - the main training loop and high-level setup code
```
Comment (wconstab) [inline README.md:24]: oh- nice- didn't know you could link this way :)

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -17,20 +17,74 @@ We actively welcome your pull requests.
...
 By contributing to `torchtitan`, you agree that your contributions will be licensed
 under the LICENSE file in the root directory of this source tree.
+
+---
+
+## Contributing Guidelines
+
+### Principles of contribution
+
+- Apply PyTorch-native training techniques.
```
Comment (wconstab) [inline CONTRIBUTING.md:44]: i wonder if we should justify more why we have this principle, or its enough to just state it...

**Review by wconstab — APPROVED** [review]
overall LGTM!

File: CONTRIBUTING.md (hunk 5 above)
Comment (tianyu-l) [reply CONTRIBUTING.md:44]: because the repo is sitting under pytorch lol?
I think this is beyond the scope of this PR. E.g. in the README first sentence, we set the tone "`torchtitan` is a proof-of-concept for Large-scale LLM training using native PyTorch." Maybe we can rethink about that.
